prover-scheduler/src/scheduler_service.rs

```diff
@@ -0,0 +1,273 @@
+            scheduler_sender,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl SchedulerService for SchedulerServiceSVC {
+    type SchedulerStreamStream =
+        Pin<Box<dyn Stream<Item = Result<SchedulerMessage, Status>> + Send + Sync + 'static>>;
+
+    async fn scheduler_stream(
+        &self,
+        request: Request<Streaming<BatchProverMessage>>,
+    ) -> Result<Response<Self::SchedulerStreamStream>, Status> {
+        let mut stream = request.into_inner();
+        let (tx, rx) = mpsc::channel(10);
+        let scheduler_sender = self.scheduler_sender.clone();
+
+        tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    Some(result) = stream.next() => {
+                        match result {
+                            Ok(batch_prover_msg) => {
+                                if let Some(msg) = batch_prover_msg.message_type {
+                                    let resp = match msg {
+                                        // update pb, we don't need too much information
+                                        batch_prover_message::MessageType::Registry(r) => {
+                                            if let Ok(schdeduler_msg) = handle_get_status_response(r, scheduler_sender.clone()).await {
+                                                schdeduler_msg
+                                            } else {
+                                                // close the connection
+                                                break;
+                                            }
+                                        }
+                                        // update pb, we don't need GeneBatchProofResponse
+                                        // just wait for the result, don't need to get again
+                                        batch_prover_message::MessageType::TakeBatchProofTask(r) => {
+                                            // TODO: id
+                                            if let Ok(scheduler_msg) = handle_gen_batch_proof_response(r.prover_id.clone(), scheduler_sender.clone()).await {
+                                                scheduler_msg
+                                            } else {
+                                                // close the connection
+                                                break;
+                                            }
+                                        }
+                                        // receive proof, trigger next batch_proof task
+                                        batch_prover_message::MessageType::BatchProofResult(r) => {
+                                            if let Ok(scheduler_msg) = handle_get_proof_response(r, scheduler_sender.clone()).await {
+                                                scheduler_msg
+                                            } else {
+                                                // close the connection
```

Comment:
If the batch prover is not alive, can we capture this and skip the break?

Reply:
Here, we choose to break when the scheduler's event channel is closed, because we can no longer communicate with the scheduler. 
you are right, I will confirm if I should try again